SalesforceFoundation/HEDAP
# TDTM Filtering not working correctly when using Filter Field of Boolean type

# Critical Changes

# Changes
Fixed an issue that prevented the Trigger Handler from filtering properly by field values in Filter Field and Filter Value. 

# Issues Closed
Fixes: #465 

# New Metadata

# Deleted Metadata

# Testing Notes
For this PR, we need to test the following scenarios for the 'Filter Field' setting:
	1.	Standard Field (such as LastName)
	2.	Custom field in HEDA package (such as Deceased__c)
	3.	Custom field from unmanaged object (such as user's custom field MyField__c)
	4.	Custom field from another package (such as third party field tfp__TFCheckBox__c)
        5.     Relationship field (such as Account.Name, Current_Address__r.Parent_Account__r.RecordType.Name)

Recommended Steps to Test this PR:
1.	Deploy the HEDA code base to a Developer org and generate a Beta package. For QA, following the normal release process to get a Beta org
2.	Generate another Beta package with a different namespace to simulate third party package (or use this link to install one https://login.salesforce.com/packaging/installPackage.apexp?p0=04t1I000001mTMo)
3.	When both HEDA and third party package get installed, make sure the fields you need are in the object's page layout
4.	Create a simple Apex class that do a simple update of the record. For example, update Contact's "Military Background" to "Please Contact VA Services"

```
global class MyTDTM extends hed.TDTM_Runnable {
    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
        hed.TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
		if(newlist != null) {
		    for (integer i=0; i<newlist.size(); i++) {
		         Contact c = (Contact)newlist[i];
		         c.hed__Military_Background__c = 'Please Contact VA Services';
		    }
		}
		return new DmlWrapper();
	}
}
```

5.	Create a Trigger Handler record for the Apex class that you created before. 
                For example: 
                Class = class created in Step 4 
                Object = Contact
                 Trigger Action = Before Insert; Before Update 
                 Filter Field = LastName     
                 Filter Value = aaa

6.	Confirmed that the Contact with last name = "aaa" will not have Military Background = "Please Contact VA Services" after you try to Save the Contact.
7.	Now, update the Filter Field and Field Value to test all 4 scenarios mentioned above with different field type (Checkbox, Picklist, Text). Note: if you install the package with the link in step2, you should have three fields on Contact available for you (tfp__TFCheckBox__c, tfp__TFPicklist__c, tfp__TFText__c).

Example Scenarios to Test this PR:
Basically this new proposal include two parts. 
1. We will implement a validation when user enter the “Filter Field” in Trigger Handler. The validation is to prevent user enter the field with namespace. 
2. We will also fix the filter logic to let user use their own field. 

Say, we have those fields on Contact object in an org:
Standard Field: LastName
HEDA Field: Deceased__c
Non HEDA Field: MyField__c 
Field from another package: tfp__TFCheckBox__c
Non HEDA Object: MyObject__c

(Filter Field Validation):
LastName : Valid
Deceased__c: Valid
hed__Deceased__c: NOT Valid
MyField__c: Valid
tfp__TFCheckBox__c: Valid
MyObject__r.MyObjectField__c Valid

(Actual Filter logic):
LastName : Works
Deceased__c: Works
hed__Deceased__c: Doesn't Work
MyField__c: Works:
tfp__TFCheckBox__c: Works
MyObject__r.MyObjectField__c Works

* Special case: Create a field in the org that has the same api name as HEDA field. For example, create a new field call Deceased in the org (the API name should be Deceased__c). Put Deceased__c into the Filter Field in the Trigger Handler. You should see an error message after you click Save. Check if the message make sense and following the message to Save the record again. 

Please note that the scenarios here only cover the checkbox field type. Please try different field type (Checkbox, Picklist, Text) as mentioned in step 7.
